README.md
# s3storagefrontend

<!---
[![start with why](https://img.shields.io/badge/start%20with-why%3F-brightgreen.svg?style=flat)](http://www.ted.com/talks/simon_sinek_how_great_leaders_inspire_action)
--->
[![GitHub release](https://img.shields.io/github/release/elbosso/s3storagefrontend/all.svg?maxAge=1)](https://GitHub.com/elbosso/s3storagefrontend/releases/)
[![GitHub tag](https://img.shields.io/github/tag/elbosso/s3storagefrontend.svg)](https://GitHub.com/elbosso/s3storagefrontend/tags/)
[![GitHub license](https://img.shields.io/github/license/elbosso/s3storagefrontend.svg)](https://github.com/elbosso/s3storagefrontend/blob/master/LICENSE)
[![GitHub issues](https://img.shields.io/github/issues/elbosso/s3storagefrontend.svg)](https://GitHub.com/elbosso/s3storagefrontend/issues/)
[![GitHub issues-closed](https://img.shields.io/github/issues-closed/elbosso/s3storagefrontend.svg)](https://GitHub.com/elbosso/s3storagefrontend/issues?q=is%3Aissue+is%3Aclosed)
[![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat)](https://github.com/elbosso/s3storagefrontend/issues)
[![GitHub contributors](https://img.shields.io/github/contributors/elbosso/s3storagefrontend.svg)](https://GitHub.com/elbosso/s3storagefrontend/graphs/contributors/)
[![Github All Releases](https://img.shields.io/github/downloads/elbosso/s3storagefrontend/total.svg)](https://github.com/elbosso/s3storagefrontend)
[![Website elbosso.github.io](https://img.shields.io/website-up-down-green-red/https/elbosso.github.io.svg)](https://elbosso.github.io/)

![s3storagefrontend_logo](src/main/resources/site/s3storagefrontend_logo.png)

## Overview

This project offers a frontend to an s3 storage - storing files only for a short time. 
You can build it by issuing

```
mvn compile package
```

and then starting the resulting monolithic jar file by issuing

```
$JAVA_HOME/bin/java -jar target/s3storagefrontend-<version>-jar-with-dependencies.jar
```

Alternatively one could just start the server using maven by  issuing

```
mvn compile exec:java
```

In both cases, the server starts on port 7000 - at the moment
only POST requests are supported. POSTs must be of
mimetype `multipart/form-data`, the form must contain a file
named _data_ containing the contents of the file being uploaded - it 
is then answered with a key needed to retrieve the file again.

At the moment, this is a prototype. It still lacks support for TLS.

However the recommended mode of using this is to use the provided _Dockerfile_ 
and _docker-compose.yml_ file. It is probably better 
to actually use a proxy solution like traefik (the docker-compose is 
already prepared for this) or similar
solutions so the services are actually accessible with a sound hostname and 
some default port.

## Configuration

All configuration is done vie _environment.env_ - just copy and rename _environment.env_template_ 
and customize its contents. Monitoring is configured by copying _src/main/resources/influxdb_micrometer.proprerties_template_
to _src/main/resources/influxdb_micrometer.proprerties_ and customizing it.

## Where to get an S3 instance for playing?

You could use [Min.IO](https://github.com/minio/minio)
for your first steps -i even prepared a small `docker-compose.yml` as stasrrting point to host your own instance:

```
version: '3.1'

services:
  app:
    image: minio/minio
    env_file: ./minio.env
    volumes:
     - ./data:/data
    container_name: minio
    hostname: minio
    labels:
      - "traefik.enable=true"
      - "traefik.http.routers.minio.rule=Host(`minio.pi-docker.lab`)"
      - "traefik.http.services.minio.loadbalancer.server.port=9000"
      - "traefik.docker.network=pi_traefik_proxy"
    networks:
      - traefik_proxy
    command:  minio server /data

networks:
  traefik_proxy:
    external:
      name: pi_traefik_proxy
```

## Working with it

### Upload

Just issue a HTTP POST request as multipart form data
(for example from a file upload from inside a web page):
```shell script
curl -J -O -F "data=@<some_file>" http://<host>:<port>/upload
``` 

A HTTP status code of 201 signifies success. If the operation is successful, the response
contains the direct link fpr the download of the uploaded object in three different locations:
The URL of the uploaded file (to be used to access the file later on or to share it with others)
is contained in the `head` of the HTML page as a `link` element,
in the body of the HTML page as a hyperlink (`a`) element and as a `Content-Location` header in the response.

If the HTTP header `Accept` is sent with the request having a value of `text/plain`,
the service does only return the plain URL to the uploaded file for easier integration into scripts for example:</p>
```shell script
curl -J -O -H "Accept: text/plain" -F "data=@<some_file_name>" http://<host>:<port>/upload 
```

If the HTTP header `Accept` is sent with the request having a value of `application/json`,
the service does return a json document with helpful information - for example:</p>
```shell script
curl -J -O -H "Accept: application/json" -F "data=@<some_file_name>" http://<host>:<port>/upload 
```

would return something on the lines of:

```json
{
  "download": {
    "curl": "curl -O -J http://localhost:7000/download/1ecd4d93-d0bb-45e3-8aff-c9bc3662f0c9",
    "wget": "wget http://localhost:7000/download/1ecd4d93-d0bb-45e3-8aff-c9bc3662f0c9 --content-disposition",
    "href": "http://localhost:7000/download/1ecd4d93-d0bb-45e3-8aff-c9bc3662f0c9"
  },
  "delete": {
    "curl": "cirl -X DELETE http://localhost:7000/delete/1ecd4d93-d0bb-45e3-8aff-c9bc3662f0c9",
    "wget": "wget --method=DELETE http://localhost:7000/delete/1ecd4d93-d0bb-45e3-8aff-c9bc3662f0c9",
    "href": "http://localhost:7000/delete/1ecd4d93-d0bb-45e3-8aff-c9bc3662f0c9"
  },
  "uuid": "1ecd4d93-d0bb-45e3-8aff-c9bc3662f0c9"
}
```

### Retrieval

Once the user has a valid URL, accessing the file is as easy as</p>
```shell script
curl -O -J http://<host>:<port>/<url></div>
```
or alternatively using
`wget:`
```shell script
wget http://<host>:<port>/<url> --content-disposition
```

### Deletion

Deleting a file explicitly before the configured timespan is up is almost as easy:
All the user needs to do that is the UUID contained in the URL:
```shell script
curl -X DELETE http://<host>:<port>/delete/<uuid>
```

## Expiration of documents

At the moment the bucket given in _environment.env_ must exist. However it is not necessary to
configure any expiration times upon it - after starting the service, the bucket is configured accordingly.

## RFC 3161 Timestamps

It is possible to configure the system in a way that allows to fetch cryptographic timestamps
for the file contents - al the administrator has to do is to specify four configuration items.
If this is done, the solution offers an additional endpoint named `timestamp` that works
similarly to the `download` endpoint but does not return the files original content but
a timestamp according to [RFC 3161](https://tools.ietf.org/html/rfc3161), making it possible to prove that the archived document
existed in its current form at the time the timestamp was created. The four configuration items needed are:

* `de.elbosso.tools.s3storagefrontend.rest.App.rfc3161url` The Url the timestamping 
  service is available under - if not given, the timestamping service is not 
  available. This is the default.
* `de.elbosso.tools.s3storagefrontend.rest.App.tspolicyoid` The policy to use for 
  the timestamping request - if not given, 
  [0.4.0.2023.1.1](http://oid-info.com/get/0.4.0.2023.1.1) is used - 
  the default baseline policy
* `de.elbosso.tools.s3storagefrontend.rest.App.tscertreq` decides wether the reply from
  the timestamping service should include the certificate - the default here is `true`
* `de.elbosso.tools.s3storagefrontend.rest.App.binarybodyname` The timestamping service
  is queried using a `POST` request with a multipart form data - this configuration item
  specifies the key under which the timestamping query is registered in the form - the
  default here is `tsq`
  
It is for example possible to use the project 
[rfc3161timestampingserver](https://github.com/elbosso/rfc3161timestampingserver)
as timestamping service here.

## Attribution
The LICENSE holds for all material in this project except for _src/main/resources/site/s3storagefrontend_logo.png_ and _src/main/resources/site/s3storagefrontend_logo.svg_ - they are made available here under the [Creative Commons Attribution-Share Alike 3.0 Unported](https://creativecommons.org/licenses/by-sa/3.0/deed.en) license. They were derived from [File:Server-multiple.svg](https://en.wikipedia.org/wiki/File:Server-multiple.svg).


CHANGELOG.md
# Changelog

Changelog of this project.


## Unreleased
    
        
            
### GitHub [#9](https://github.com/elbosso/s3storagefrontend/issues/9) get rid of log4j and use the new bom dependency
            
            
        
        

        
**fixes #9**

            

[4ae00c84b485c9b](https://github.com/elbosso/s3storagefrontend/commit/4ae00c84b485c9b) Juergen Key *2022-04-27 15:55:24*

  

 
        
        
### No issue
        

        
**[maven-release-plugin] prepare for next development iteration**

            

[a00df87c7fb9e84](https://github.com/elbosso/s3storagefrontend/commit/a00df87c7fb9e84) Juergen Key *2022-04-27 15:55:10*

  

 

## 1.4.0
    
        
            
### GitHub [#9](https://github.com/elbosso/s3storagefrontend/issues/9) get rid of log4j and use the new bom dependency
            
            
        
        

        
**re #9**

            

[7e7f5817ac8e2c2](https://github.com/elbosso/s3storagefrontend/commit/7e7f5817ac8e2c2) Juergen Key *2022-04-27 15:54:44*

  

 
        
        
### No issue
        

        
**[maven-release-plugin] prepare release s3storagefrontend-1.4.0**

            

[744cbf39c66c8f3](https://github.com/elbosso/s3storagefrontend/commit/744cbf39c66c8f3) Juergen Key *2022-04-27 15:55:07*

  
**change of repository location for de.elbosso artifacts**

            

[72250d7c632d1c7](https://github.com/elbosso/s3storagefrontend/commit/72250d7c632d1c7) Jürgen Key *2021-09-12 08:49:22*

  
**Merge branch 'master' of https://github.com/elbosso/s3storagefrontend**

            

[b3efb3dff97877b](https://github.com/elbosso/s3storagefrontend/commit/b3efb3dff97877b) Juergen Key *2021-05-08 06:27:01*

  
**added https**

            

[f3ed330a7190ed3](https://github.com/elbosso/s3storagefrontend/commit/f3ed330a7190ed3) Juergen Key *2021-05-08 06:26:35*

  
**logo tweaked**

            

[bdc4d2ea74db726](https://github.com/elbosso/s3storagefrontend/commit/bdc4d2ea74db726) Jürgen Key *2021-03-22 15:51:37*

  
**logo added to the landing page**

            

[6499ed5f45d6363](https://github.com/elbosso/s3storagefrontend/commit/6499ed5f45d6363) Jürgen Key *2021-03-22 07:06:49*

  
**attribution for logos**

            

[5630bcda98c2086](https://github.com/elbosso/s3storagefrontend/commit/5630bcda98c2086) Jürgen Key *2021-03-21 18:35:06*

  
**added logo**

            

[e9bd02fd81b06ac](https://github.com/elbosso/s3storagefrontend/commit/e9bd02fd81b06ac) Jürgen Key *2021-03-21 18:31:04*

  
**added logo**

            

[4484bd29d70293a](https://github.com/elbosso/s3storagefrontend/commit/4484bd29d70293a) Jürgen Key *2021-03-21 18:30:32*

  
**Update README.md**

            
* added link to rfc 3161
            

[58b7a00e5876ef2](https://github.com/elbosso/s3storagefrontend/commit/58b7a00e5876ef2) Jürgen Key *2021-03-21 16:10:38*

  
**housekeeping**

            

[08d3aa05b9f441f](https://github.com/elbosso/s3storagefrontend/commit/08d3aa05b9f441f) Jürgen Key *2021-03-21 10:02:28*

  
**[maven-release-plugin] prepare for next development iteration**

            

[4731ce962f58d72](https://github.com/elbosso/s3storagefrontend/commit/4731ce962f58d72) Jürgen Key *2021-03-21 10:01:48*

  
**release fxxx-up**

            

[e88693810f06ffb](https://github.com/elbosso/s3storagefrontend/commit/e88693810f06ffb) Jürgen Key *2021-03-20 09:45:40*

  
**[maven-release-plugin] prepare for next development iteration**

            

[e32706dd434b487](https://github.com/elbosso/s3storagefrontend/commit/e32706dd434b487) Jürgen Key *2021-03-20 09:45:05*

  

 

## 1.3.0
    
        
            
### GitHub [#6](https://github.com/elbosso/s3storagefrontend/issues/6) possibility to specify &quot;accept&quot; header as application/json
            
            
        
        

        
**fixes #6**

            

[cd15e2f05b4e4d7](https://github.com/elbosso/s3storagefrontend/commit/cd15e2f05b4e4d7) Jürgen Key *2021-03-20 14:35:27*

  
**re #6**

            

[ee0d8febbad1100](https://github.com/elbosso/s3storagefrontend/commit/ee0d8febbad1100) Jürgen Key *2021-03-20 11:45:23*

  

 
        
            
### GitHub [#7](https://github.com/elbosso/s3storagefrontend/issues/7) add possibility to automatically create a cryptographic timestamp (RFC3161) for an uploaded file
            
            
        
        

        
**fixes #7**

            

[d68087d30c577cb](https://github.com/elbosso/s3storagefrontend/commit/d68087d30c577cb) Jürgen Key *2021-03-21 10:00:35*

  
**re #7**

            

[4ce4e14d40f75c5](https://github.com/elbosso/s3storagefrontend/commit/4ce4e14d40f75c5) Jürgen Key *2021-03-21 09:48:48*

  
**re #7**

            

[68f325f1e592189](https://github.com/elbosso/s3storagefrontend/commit/68f325f1e592189) Jürgen Key *2021-03-21 08:21:20*

  
**re #7**

            

[703a46a8eb9850b](https://github.com/elbosso/s3storagefrontend/commit/703a46a8eb9850b) Jürgen Key *2021-03-21 07:26:34*

  

 
        
            
### GitHub [#8](https://github.com/elbosso/s3storagefrontend/issues/8) make the duration files are kept for configurable with a default of one day
            
            
        
        

        
**re #8**

            

[1b410734d3dbfd4](https://github.com/elbosso/s3storagefrontend/commit/1b410734d3dbfd4) Jürgen Key *2021-03-20 17:52:20*

  
**re #8**

            

[ffd6d451d60a76e](https://github.com/elbosso/s3storagefrontend/commit/ffd6d451d60a76e) Jürgen Key *2021-03-20 17:50:03*

  
**re #8**

            

[80a34edc0f00c4d](https://github.com/elbosso/s3storagefrontend/commit/80a34edc0f00c4d) Jürgen Key *2021-03-20 17:46:33*

  
**fixes #8**

            

[8f4274fac96f94e](https://github.com/elbosso/s3storagefrontend/commit/8f4274fac96f94e) Jürgen Key *2021-03-20 17:45:44*

  

 
        
        
### No issue
        

        
**[maven-release-plugin] prepare release s3storagefrontend-1.3.0**

            


src/main/java/de/elbosso/tools/s3storagefrontend/rest/App.java
package de.elbosso.tools.s3storagefrontend.rest;

import ch.qos.logback.classic.Level;
import com.amazonaws.client.builder.AwsClientBuilder;
import com.amazonaws.regions.Regions;
import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.AmazonS3ClientBuilder;
import com.amazonaws.services.s3.model.BucketLifecycleConfiguration;
import de.elbosso.tools.s3storagefrontend.rest.handlers.DeletionHandler;
import de.elbosso.tools.s3storagefrontend.rest.handlers.DownloadHandler;
import de.elbosso.tools.s3storagefrontend.rest.handlers.Rfc3161Handler;
import de.elbosso.tools.s3storagefrontend.rest.handlers.UploadHandler;
import io.javalin.Javalin;
import io.micrometer.core.instrument.Clock;
import io.micrometer.core.instrument.MeterRegistry;
import io.micrometer.core.instrument.Metrics;
import io.micrometer.influx.InfluxMeterRegistry;

import java.util.Arrays;

public class App {
	private final static java.lang.String OLDRULEID="delete rule";
	private final static java.lang.String RULEID="de.elbosso.tools.s3storagefrontend.rest.App.RULEID";
	private static final java.lang.String EXPIRATIONENVKEY="de.elbosso.tools.s3storagefrontend.rest.App.expirationInDays";
	private final static org.slf4j.Logger CLASS_LOGGER=org.slf4j.LoggerFactory.getLogger(App.class);
	private final static org.slf4j.Logger EXCEPTION_LOGGER=org.slf4j.LoggerFactory.getLogger("ExceptionCatcher");

	public static void main(String[] args)
	{
		de.elbosso.util.Utilities.configureBasicStdoutLogging(Level.ALL);
		io.micrometer.influx.InfluxConfig config = new de.elbosso.tools.s3storagefrontend.monitoring.InfluxConfig();
		MeterRegistry registry = new InfluxMeterRegistry(config, Clock.SYSTEM);
		Metrics.globalRegistry.add(registry);
		App.init();
	}
	private static final Javalin init()
	{
		Regions clientRegion = Regions.DEFAULT_REGION;
		AmazonS3 s3Client = AmazonS3ClientBuilder.standard()
//							.withRegion(clientRegion)
				.withEndpointConfiguration(new AwsClientBuilder.EndpointConfiguration(System.getenv("AWS_S3_URL"), clientRegion.name()))
				.withPathStyleAccessEnabled(true)
//							.withClientConfiguration(clientConfiguration)
				.build();
		String bucketName = System.getenv("AWS_BUCKET_NAME");

		BucketLifecycleConfiguration blcc=s3Client.getBucketLifecycleConfiguration(bucketName);
		//get Rules for Bucket
		java.util.List<BucketLifecycleConfiguration.Rule> rules=blcc.getRules();
		if(rules==null)
			rules=new java.util.LinkedList();
		BucketLifecycleConfiguration.Rule expirationRule = null;
		BucketLifecycleConfiguration.Rule oldRule = null;
		for(BucketLifecycleConfiguration.Rule rule:rules)
		{
			if(CLASS_LOGGER.isDebugEnabled())CLASS_LOGGER.debug(rule.getId());
			if(CLASS_LOGGER.isDebugEnabled())CLASS_LOGGER.debug(java.util.Objects.toString(rule.getExpirationDate()));
			if(CLASS_LOGGER.isDebugEnabled())CLASS_LOGGER.debug(java.util.Objects.toString(rule.getExpirationInDays()));
			if(RULEID.equals(rule.getId()))
			{
				expirationRule=rule;
			}
			else if(OLDRULEID.equals(rule.getId()))
			{
				oldRule=rule;
			}
		}
		if(oldRule!=null)
			rules.remove(oldRule);
		if(expirationRule!=null)
			rules.remove(expirationRule);

		int expi=1;
		java.lang.String expirationInDays=System.getenv(EXPIRATIONENVKEY);
		if(expirationInDays!=null)
		{
			try
			{
				expi = java.lang.Integer.parseInt(expirationInDays);
			}
			catch(java.lang.NumberFormatException exp)
			{
				if(CLASS_LOGGER.isDebugEnabled())CLASS_LOGGER.debug("could not parse value of "+EXPIRATIONENVKEY+" ("+expirationInDays+") - using the default value 1!");
				expi=1;
			}
		}
		if(CLASS_LOGGER.isDebugEnabled())CLASS_LOGGER.debug(EXPIRATIONENVKEY+" not found - using the default value 1!");
		if(expi<1)
		{
			if(CLASS_LOGGER.isDebugEnabled())CLASS_LOGGER.debug("values of less than one day are not allowed for the time until expiration - using the default value of 1!");
			expi=1;
		}
		// Create a rule to delete objects after the specified number of days.
		// The rule applies to all objects with the tag "archive" set to "true".
		expirationRule = new BucketLifecycleConfiguration.Rule()
				.withId(RULEID)
				.withExpirationInDays(expi)
				.withStatus(BucketLifecycleConfiguration.ENABLED);
		rules.add(expirationRule);

		// Add the rules to a new BucketLifecycleConfiguration.
		blcc = new BucketLifecycleConfiguration().withRules(rules);

		// Save the configuration.
		s3Client.setBucketLifecycleConfiguration(bucketName, blcc);

		Javalin app = Javalin.create().start(7000);
		if(CLASS_LOGGER.isDebugEnabled())CLASS_LOGGER.debug("started app - listening on port 7000");
		app.config.addStaticFiles("/site");
		if(CLASS_LOGGER.isDebugEnabled())CLASS_LOGGER.debug("added path for static contents: /site (allowed methods: GET)");
		DownloadHandler.register(app);
		UploadHandler.register(app);
		DeletionHandler.register(app);
		Rfc3161Handler.register(app);
		app.before(ctx -> {
			if(CLASS_LOGGER.isDebugEnabled())CLASS_LOGGER.debug(ctx.req.getMethod()+" "+ctx.contentType());
		});
		if(CLASS_LOGGER.isDebugEnabled())CLASS_LOGGER.debug("added before interceptor");
		Runtime.getRuntime().addShutdownHook(new Thread(() -> {
			app.stop();
		}));
		if(CLASS_LOGGER.isDebugEnabled())CLASS_LOGGER.debug("added callback for stopping the application");
		app.events(event -> {
			event.serverStopping(() -> { /* Your code here */ });
			event.serverStopped(() -> {
				if(CLASS_LOGGER.isDebugEnabled())CLASS_LOGGER.debug("Server stopped");
			});
			if(CLASS_LOGGER.isDebugEnabled())CLASS_LOGGER.debug("added listener for server stopped event");
		});
		return app;
	}
}


src/main/java/de/elbosso/tools/s3storagefrontend/rest/handlers/UploadHandler.java
package de.elbosso.tools.s3storagefrontend.rest.handlers;

import com.amazonaws.AmazonClientException;
import com.amazonaws.AmazonServiceException;
import com.amazonaws.client.builder.AwsClientBuilder;
import com.amazonaws.regions.Regions;
import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.AmazonS3ClientBuilder;
import com.amazonaws.services.s3.model.*;
import io.javalin.Javalin;
import io.javalin.http.Context;
import io.javalin.http.Handler;
import io.micrometer.core.instrument.Metrics;
import java.util.UUID;

public class UploadHandler extends Object implements Handler
{
	final static java.lang.String RESOURCENAME="upload";
	private final static org.slf4j.Logger CLASS_LOGGER=org.slf4j.LoggerFactory.getLogger(UploadHandler.class);
	private final static org.slf4j.Logger EXCEPTION_LOGGER=org.slf4j.LoggerFactory.getLogger("ExceptionCatcher");

	public static void register(Javalin app)
	{
		app.post("/"+RESOURCENAME, new UploadHandler());
		if(CLASS_LOGGER.isDebugEnabled())CLASS_LOGGER.debug("added path for storing data: /"+RESOURCENAME+" (allowed methods: POST)");
	}
	@Override
	public void handle(Context ctx) throws Exception
	{
		if(CLASS_LOGGER.isDebugEnabled())CLASS_LOGGER.debug("received upload request");
		byte[] data=null;
		java.lang.String accepts=ctx.header("Accept");
		if(accepts!=null)
			if(CLASS_LOGGER.isDebugEnabled())CLASS_LOGGER.debug("request specified wanted return as "+accepts);
		java.lang.String contentType=ctx.contentType();
		String fileObjKeyName = UUID.randomUUID().toString();
		java.lang.String s3ContentDisposition=fileObjKeyName;
		java.lang.String s3ContentType="application/octet-stream";
		if(ctx.contentType().startsWith("multipart/form-data"))
		{
			if(CLASS_LOGGER.isDebugEnabled())CLASS_LOGGER.debug("request is multipart/form-data - searching for data under key\"data\"");
			//curl -F "data=@somefile.suffix" http://localhost:7000/upload
			if(ctx.uploadedFile("data")!=null)
			{
				if(CLASS_LOGGER.isDebugEnabled())CLASS_LOGGER.debug("found it - data length is "+ctx.uploadedFile("data").getContentLength());
				s3ContentDisposition=ctx.uploadedFile("data").getFilename();
				s3ContentType=ctx.uploadedFile("data").getContentType();
				java.io.InputStream is=ctx.uploadedFile("data").getContent();
				java.io.ByteArrayOutputStream baos=new java.io.ByteArrayOutputStream();
				de.elbosso.util.Utilities.copyBetweenStreams(is,baos,true);
				data=baos.toByteArray();
			}
			else
			{
				if(CLASS_LOGGER.isErrorEnabled())CLASS_LOGGER.error("no field named \"data\" found in form data . corrupted request?");
				Metrics.counter("s3storagefrontend.post", "resourcename","/"+RESOURCENAME,"httpstatus","500","error","data not found","contentType",contentType,"remoteAddr",ctx.req.getRemoteAddr(),"remoteHost",ctx.req.getRemoteHost(),"localAddr",ctx.req.getLocalAddr(),"localName",ctx.req.getLocalName()).increment();
			}
			if(ctx.formParam("s3ContentDisposition")!=null)
			{
				s3ContentDisposition=ctx.formParam("s3ContentDisposition");
				if (CLASS_LOGGER.isDebugEnabled()) CLASS_LOGGER.debug("found s3ContentDisposition to be " + s3ContentDisposition);
			}
			else
			{
				if (CLASS_LOGGER.isWarnEnabled()) CLASS_LOGGER.warn("did not find s3ContentDisposition");
			}
			if(ctx.formParam("s3ContentType")!=null)
			{
				s3ContentType = ctx.formParam("s3ContentType");
				if (CLASS_LOGGER.isDebugEnabled()) CLASS_LOGGER.debug("found s3ContentType to be " + s3ContentType);
			}
			else
			{
				if (CLASS_LOGGER.isWarnEnabled()) CLASS_LOGGER.warn("did not find s3ContentType");
			}
		}
		else
		{
			if(CLASS_LOGGER.isErrorEnabled())CLASS_LOGGER.error("request is not multipart/form-data ");
			Metrics.counter("s3storagefrontend.post", "resourcename","/"+RESOURCENAME,"httpstatus","500","error","encoding","contentType",contentType,"remoteAddr",ctx.req.getRemoteAddr(),"remoteHost",ctx.req.getRemoteHost(),"localAddr",ctx.req.getLocalAddr(),"localName",ctx.req.getLocalName()).increment();
		}
		if(data!=null)
		{
			try
			{
				Regions clientRegion = Regions.DEFAULT_REGION;
				AmazonS3 s3Client = AmazonS3ClientBuilder.standard()
//							.withRegion(clientRegion)
						.withEndpointConfiguration(new AwsClientBuilder.EndpointConfiguration(System.getenv("AWS_S3_URL"), clientRegion.name()))
						.withPathStyleAccessEnabled(true)
//							.withClientConfiguration(clientConfiguration)
						.build();
				String bucketName = System.getenv("AWS_BUCKET_NAME");

				java.io.ByteArrayInputStream bais = new java.io.ByteArrayInputStream(data);
				// Upload a file as a new object with ContentType and title specified.
				ObjectMetadata metadata = new ObjectMetadata();
				metadata.setContentDisposition(s3ContentDisposition);
				metadata.setContentType(s3ContentType);
				metadata.setContentLength(data.length);

				//					metadata.addUserMetadata("title", "someTitle");
				PutObjectRequest request = new PutObjectRequest(bucketName, fileObjKeyName, bais,metadata);
				//https://docs.aws.amazon.com/AmazonS3/latest/userguide/tagging-managing.html
				java.util.List<Tag> tags = new java.util.ArrayList<Tag>();
				tags.add(new Tag("archive", "true"));
				request.setTagging(new ObjectTagging(tags));
				s3Client.putObject(request);
				ctx.status(201);
				//ctx.header("Content-Disposition","filename=\"reply.tsr\"");
				//ctx.result(new java.io.ByteArrayInputStream(tsr));
				java.lang.String downloadHref=ctx.fullUrl().substring(0,ctx.fullUrl().length()-RESOURCENAME.length())+DownloadHandler.RESOURCENAME+"/"+fileObjKeyName;
				java.lang.String deleteHref=ctx.fullUrl().substring(0,ctx.fullUrl().length()-RESOURCENAME.length())+DeletionHandler.RESOURCENAME+"/"+fileObjKeyName;
				java.lang.String timestampHref=ctx.fullUrl().substring(0,ctx.fullUrl().length()-RESOURCENAME.length())+Rfc3161Handler.RESOURCENAME+"/"+fileObjKeyName;
				ctx.header("Content-Location",downloadHref);
				if(accepts.equals("text/plain"))
				{
					ctx.contentType(accepts);
					String cd=s3ContentDisposition.substring(0,s3ContentDisposition.lastIndexOf("."))+".txt";
					ctx.header("Content-Disposition", "filename=\"" + cd + "\"");
					ctx.result(downloadHref);
				}
				else if(accepts.equals("application/json"))
				{
					ctx.contentType("application/json");
					java.util.Map<java.lang.String,java.lang.Object> map=new java.util.HashMap();
					java.util.Map<java.lang.String,java.lang.String> downloadmap=new java.util.HashMap();
					java.util.Map<java.lang.String,java.lang.String> deletemap=new java.util.HashMap();
					map.put("download",downloadmap);
					downloadmap.put("href",downloadHref);
					downloadmap.put("wget","wget "+downloadHref+" --content-disposition");
					downloadmap.put("curl","curl -O -J "+downloadHref);
					map.put("delete",deletemap);
					deletemap.put("href",deleteHref);
					deletemap.put("wget","wget --method=DELETE "+deleteHref);
					deletemap.put("curl","curl -X DELETE "+deleteHref);
					map.put("uuid",fileObjKeyName);
					if(System.getenv(Rfc3161Handler.RFC3161URLENVKEY)!=null)
					{
						java.util.Map<java.lang.String,java.lang.String> timestampmap=new java.util.HashMap();
						timestampmap.put("href",timestampHref);
						timestampmap.put("wget","wget "+timestampHref+" --content-disposition");
						timestampmap.put("curl","curl -O -J "+timestampHref);
						map.put("timestamp",timestampmap);
					}
					String cd=s3ContentDisposition.substring(0,s3ContentDisposition.lastIndexOf("."))+".json";
					ctx.header("Content-Disposition", "filename=\"" + cd + "\"");
					ctx.json(map);
				}
				else
				{
					ctx.contentType("text/html");
					String cd=s3ContentDisposition.substring(0,s3ContentDisposition.lastIndexOf("."))+".html";
					ctx.header("Content-Disposition", "filename=\"" + cd + "\"");
					ctx.result("<html><head>" +
							"<link href=\"" + downloadHref + "\" rel=\"item\" type=\"" + metadata.getContentType() + "\" />" +
							"</head><body>" +
							"<a href=\"" + downloadHref + "\">Download/Share Document Content</a>" +
							(System.getenv(Rfc3161Handler.RFC3161URLENVKEY)!=null?
							"<br><a href=\"" + timestampHref + "\">Download/Share Timestamp</a>"
							:"")+
							"</body></html>");
				}
				Metrics.counter("s3storagefrontend.post", "resourcename","/"+RESOURCENAME,"httpstatus",java.lang.Integer.toString(ctx.status()),"success","true","contentType",contentType,"remoteAddr",ctx.req.getRemoteAddr(),"remoteHost",ctx.req.getRemoteHost(),"localAddr",ctx.req.getLocalAddr(),"localName",ctx.req.getLocalName()).increment();
			}
			catch (AmazonServiceException ase)
			{
				ctx.status(500);
				ctx.result(ase.getMessage());
				if(CLASS_LOGGER.isErrorEnabled())CLASS_LOGGER.error("Caught an AmazonServiceException, which " + "means your request made it "
						+ "to Amazon S3, but was rejected with an error response" + " for some reason.");
				if(CLASS_LOGGER.isErrorEnabled())CLASS_LOGGER.error("Error Message:    " + ase.getMessage());
				if(CLASS_LOGGER.isErrorEnabled())CLASS_LOGGER.error("HTTP Status Code: " + ase.getStatusCode());
				if(CLASS_LOGGER.isErrorEnabled())CLASS_LOGGER.error("AWS Error Code:   " + ase.getErrorCode());
				if(CLASS_LOGGER.isErrorEnabled())CLASS_LOGGER.error("Error Type:       " + ase.getErrorType());
				if(CLASS_LOGGER.isErrorEnabled())CLASS_LOGGER.error("Request ID:       " + ase.getRequestId());
				if(EXCEPTION_LOGGER.isErrorEnabled())EXCEPTION_LOGGER.error(ase.getMessage(),ase);
				Metrics.counter("s3storagefrontend.post", "resourcename","/"+RESOURCENAME,"httpstatus",java.lang.Integer.toString(ctx.status()),"error",(ase.getMessage()!=null?ase.getMessage():"NPE"),"contentType",contentType,"remoteAddr",ctx.req.getRemoteAddr(),"remoteHost",ctx.req.getRemoteHost(),"localAddr",ctx.req.getLocalAddr(),"localName",ctx.req.getLocalName()).increment();
			}
			catch (AmazonClientException ace)
			{
				ctx.status(500);
				ctx.result(ace.getMessage());
				if(CLASS_LOGGER.isErrorEnabled())CLASS_LOGGER.error("Caught an AmazonClientException, which " + "means the client encountered " + "an internal error while trying to "
						+ "communicate with S3, " + "such as not being able to access the network.");
				if(EXCEPTION_LOGGER.isErrorEnabled())EXCEPTION_LOGGER.error(ace.getMessage(),ace);
				Metrics.counter("s3storagefrontend.post", "resourcename","/"+RESOURCENAME,"httpstatus",java.lang.Integer.toString(ctx.status()),"error",(ace.getMessage()!=null?ace.getMessage():"NPE"),"contentType",contentType,"remoteAddr",ctx.req.getRemoteAddr(),"remoteHost",ctx.req.getRemoteHost(),"localAddr",ctx.req.getLocalAddr(),"localName",ctx.req.getLocalName()).increment();
			}
			catch(java.lang.Throwable t)
			{
				ctx.status(500);
				ctx.result(t.getMessage());
				EXCEPTION_LOGGER.error(t.getMessage(),t);
				Metrics.counter("s3storagefrontend.post", "resourcename","/"+RESOURCENAME,"httpstatus",java.lang.Integer.toString(ctx.status()),"error",(t.getMessage()!=null?t.getMessage():"NPE"),"contentType",contentType,"remoteAddr",ctx.req.getRemoteAddr(),"remoteHost",ctx.req.getRemoteHost(),"localAddr",ctx.req.getLocalAddr(),"localName",ctx.req.getLocalName()).increment();
			}
			finally
			{
			}
		}
		else
		{
			ctx.status(500);
			ctx.result("general error");
		}
	}
}
